### ISSUE TITLE
Dependencies consumed in an in-repo-addon violate ESLint node/no-extraneous-require rule

### ISSUE THREAD

**Issue by jelhan** [ember-cli/ember-cli#8740]
Require a dependency defined in application's `package.json` in an in-repo-addon violates ESLint rule `node/no-extraneous-require`. Seems like `eslint-plugin-node` gets confuse cause in-repo-addons have a `package.json` but doesn't use it to define dependencies.

## Reproduction

Prepare a new application having an in-repo-addon using `ember-cli@3.10.1`:

```
ember new ember-cli-in-repo-addon-eslint-no-extraneous-require
cd ember-cli-in-repo-addon-eslint-no-extraneous-require
ember g in-repo-addon foo
yarn add -D fs-extra
echo "require('fs-extra');" >> lib/foo/index.js
```

Run ESLint without altering configuration:

```
$ yarn lint:js
yarn run v1.16.0
$ eslint .

/home/main/tmp/ember-cli-in-repo-addon-eslint-no-extraneous-require/lib/foo/index.js
  10:9  error  "fs-extra" is extraneous  node/no-extraneous-require

✖ 1 problem (1 error, 0 warnings)

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

```
$ ember version
ember-cli: 3.10.1
node: 8.16.0
os: linux x64
```

## Work-a-round

```patch
diff --git a/.eslintrc.js b/.eslintrc.js
index 5b13f53..2006ed1 100644
--- a/.eslintrc.js
+++ b/.eslintrc.js
@@ -45,6 +45,15 @@ module.exports = {
         // https://github.com/mysticatea/eslint-plugin-node/issues/77
         'node/no-unpublished-require': 'off'
       })
-    }
+    },
+    {
+      files: [
+        'lib/*/index.js'
+      ],
+      rules: Object.assign({}, require('eslint-plugin-node').configs.recommended.rules, {
+        // https://github.com/ember-cli/ember-cli/issues/8740
+        'node/no-extraneous-require': 'off'
+      })
+    },
   ]
 };
```

**Comment by rwjblue** [comment]
> Require a dependency defined in application's package.json in an in-repo-addon violates ESLint rule node/no-extraneous-require. Seems like eslint-plugin-node gets confuse cause in-repo-addons have a package.json but doesn't use it to define dependencies.

IMHO, the lint rule is **exactly** correct. Any dependencies used by an in-repo-addon should declare the dependency in its own `package.json`.

Most often this is done like:

```json
"dependencies": {
  "some-dep": "*"
}
```

It still relies on the top level `package.json` for actual installation of the correct version (unless you are using a monorepo or something), but it keeps it very clear that the in-repo-addon has this as a dependency.

**Comment by kellyselden** [comment]
I think we could update the blueprints to include those * deps to satisfy the linter.

**Comment by rwjblue** [comment]
@kellyselden - Not sure what you mean? AFAICT, none of these are from things added by the blueprint. In other words, the app developer has added code to `lib/some-addon/index.js` which `require`'s some package that they know is present (from the top level `package.json`), I can't see how we can detect that (other than this eslint rule) to fix it for them.

**Comment by kellyselden** [comment]
@rwjblue I was mistaken. I was thinking of the case of https://github.com/ember-cli/ember-cli/blob/master/blueprints/http-proxy/files/server/proxies/__name__.js#L8 http-proxy being introduced, with no entry being added to package.json on generate. But, separate issue.

**Comment by jelhan** [comment]
> IMHO, the lint rule is **exactly** correct. Any dependencies used by an in-repo-addon should declare the dependency in its own `package.json`.

Oh, than I got confused by the limited information available about dependencies of in-repo-addons. It's one of the points where official documentation is missing. :disappointed: Trying to Google the problem I found [this StackOverflow question](https://stackoverflow.com/questions/42353229/creating-an-ember-cli-in-app-in-repo-addon-how-to-install-npm-dependencies), which made me think that in-repo-addon package.json is not used for dependency management at all.

Thanks for the quick reply. Going to close this one now.
